### PR TITLE
check_http: add --status

### PR DESCRIPTION
**Pull request by Al2Klimov**

### CODE REVIEW

**Comment by Al2Klimov** [comment]
```
➜  monitoring-plugins git:(feature/http-status) plugins/check_http -H 127.0.0.1 -p 8080 -u /404
HTTP WARNING: HTTP/1.1 404 Not Found - 303 bytes in 0,004 second response time |time=0,003679s;;;0,000000;10,000000 size=303B;;;0
➜  monitoring-plugins git:(feature/http-status) plugins/check_http -H 127.0.0.1 -p 8080 -u /404 --status 404
HTTP OK: HTTP/1.1 404 Not Found - 303 bytes in 0,004 second response time |time=0,004444s;;;0,000000;10,000000 size=303B;;;0
➜  monitoring-plugins git:(feature/http-status) plugins/check_http -H 127.0.0.1 -p 8080 -u /404 --status 404 --status 403
HTTP OK: HTTP/1.1 404 Not Found - 303 bytes in 0,007 second response time |time=0,007177s;;;0,000000;10,000000 size=303B;;;0
➜  monitoring-plugins git:(feature/http-status)
```

**Comment by Al2Klimov** [comment]
Useful for monitoring e.g. 401s where they're required.

**Comment by sni** [comment]
Why don't you use `-e`?

**Comment by Al2Klimov** [comment]
1. IMAO a bit cheaty
2. IMAO the status check should not silently disable the (default) check for "HTTP/1."

File: plugins/check_http.c [hunk 1]
```diff
@@ -134,6 +136,33 @@ char buffer[MAX_INPUT_BUFFER];
 char *client_cert = NULL;
 char *client_privkey = NULL;
 
+/*
+ Bitmask of the status codes to yield OK on.
+ 1st bit for 100, 2nd for 101, etc.. None set means defaults.
```
Comment (RincewindsHat) [inline plugins/check_http.c:141]: Why?

File: plugins/check_http.c [hunk 2]
```diff
@@ -134,6 +136,33 @@ char buffer[MAX_INPUT_BUFFER];
...
+{
+  char status_expect_none[sizeof(status_expect)] = { 0 };
+  return !memcmp(&status_expect, &status_expect_none, sizeof(status_expect));
+}
+
+static inline
+void status_expect_set(int code)
+{
+  uint64_t idx = code - 100;
+  status_expect[idx / 64u] |= 1u << (idx % 64u);
```
Comment (RincewindsHat) [inline plugins/check_http.c:156]: Please add a comment explaining what this does or rewrite it in an understandable manner.

File: plugins/check_http.c [hunk 3]
```diff
@@ -134,6 +136,33 @@ char buffer[MAX_INPUT_BUFFER];
...
+}
+
+static inline
+int status_expect_isset(int code)
+{
+  uint64_t idx = code - 100;
+  return !!(status_expect[idx / 64u] & (1u << (idx % 64u)));
```
Comment (RincewindsHat) [inline plugins/check_http.c:163]: Please add a comment explaining what this does or rewrite it in an understandable manner.

**Review by RincewindsHat — CHANGES_REQUESTED** [review]

**Comment by RincewindsHat** [comment]
Closing this. Pls reopen, if continuation is wished for

**Pull request closed without merging**